src/charm.py

```diff
@@ -130,44 +131,41 @@ def _on_install(self, event: InstallEvent) -> None:
             self.unit.status = BlockedStatus("unable to install ZooKeeper Snap")
 
         # don't complete install until passwords set
-        if not self.peer_relation:
+        if not self.state.peer_relation:
             self.unit.status = WaitingStatus("waiting for peer relation")
             event.defer()
             return
 
-        self.set_passwords()
+        if self.unit.is_leader() and not self.state.cluster.internal_user_credentials:
+            for user in CHARM_USERS:
+                self.state.cluster.update({f"{user}-password": generate_password()})
 
         # refreshing unit hostname relation data in case ip changed
-        self.unit_peer_data.update(self.cluster.get_hostname_mapping())
+        self.state.unit_server.update(self.quorum_manager.get_hostname_mapping())
 
         # give the leader a default quorum during cluster initialisation
         if self.unit.is_leader():
-            self.app_peer_data.update({"quorum": "default - non-ssl"})
+            self.state.cluster.update({"quorum": "default - non-ssl"})
 
     def _on_cluster_relation_changed(self, event: EventBase) -> None:
```

Comment:
I wanted to keep 'standard' charm init events (`install`, `relation-joined/changed/broken`, `restart`, `start`/`pebble-ready`), in `charm.py`.
Reason being, it's the first place anyone looks when eye-balling the charm, so having it laid out here is most expected. IMO we should keep these separate, rather than having an empty `charm.py` that just uses some redundant `events/charm.py` instead.

Reply:
And on your 'should be the same' point, actually it's in this file that the biggest divergences happen between K8s + VM. I can't think of anything off the top of my head, but I recall that we handle `pebble-ready` **SLIGHTLY** different to `start`.

I do not expect that we will be able to 'copy-paste' this file, so having it centralised really helps.